File: test/molecule.cpp
#include <iterator>
#include <Helium/hemol.h>
#include <Helium/smartmol.h>
#include <Helium/concepts.h>


#include "test.h"

using namespace Helium;

void test_element_predicate()
{
  // ElementPredicate
  HeMol mol = hemol_from_smiles("C");
  ASSERT( molecule_has_atom(mol, element_eq_predicate(mol, 6)));
  ASSERT(!molecule_has_atom(mol, element_eq_predicate(mol, 7)));
  ASSERT( molecule_has_atom(mol, element_lt_predicate(mol, 7)));
  ASSERT(!molecule_has_atom(mol, element_lt_predicate(mol, 6)));
  ASSERT( molecule_has_atom(mol, element_gt_predicate(mol, 5)));
  ASSERT(!molecule_has_atom(mol, element_gt_predicate(mol, 6)));
  ASSERT( molecule_has_atom(mol, element_leq_predicate(mol, 6)));
  ASSERT( molecule_has_atom(mol, element_leq_predicate(mol, 7)));
  ASSERT(!molecule_has_atom(mol, element_leq_predicate(mol, 5)));
  ASSERT( molecule_has_atom(mol, element_geq_predicate(mol, 5)));
  ASSERT( molecule_has_atom(mol, element_geq_predicate(mol, 6)));
  ASSERT(!molecule_has_atom(mol, element_geq_predicate(mol, 7)));
}

void test_mass_predicate()
{
  // MassPredicate
  HeMol mol = hemol_from_smiles("[13C]");
  ASSERT( molecule_has_atom(mol, mass_eq_predicate(mol, 13)));
  ASSERT(!molecule_has_atom(mol, mass_eq_predicate(mol, 12)));
  ASSERT( molecule_has_atom(mol, mass_lt_predicate(mol, 14)));
  ASSERT(!molecule_has_atom(mol, mass_lt_predicate(mol, 13)));
  ASSERT( molecule_has_atom(mol, mass_gt_predicate(mol, 12)));
  ASSERT(!molecule_has_atom(mol, mass_gt_predicate(mol, 13)));
  ASSERT( molecule_has_atom(mol, mass_leq_predicate(mol, 13)));
  ASSERT(!molecule_has_atom(mol, mass_leq_predicate(mol, 12)));
  ASSERT( molecule_has_atom(mol, mass_geq_predicate(mol, 13)));
  ASSERT(!molecule_has_atom(mol, mass_geq_predicate(mol, 14)));
}

void test_charge_predicate()
{
  // ChargePredicate
  HeMol mol = hemol_from_smiles("C[NH3+]");
  ASSERT( molecule_has_atom(mol, charge_eq_predicate(mol, 1)));
  ASSERT( molecule_has_atom(mol, charge_eq_predicate(mol, 0)));
  ASSERT(!molecule_has_atom(mol, charge_eq_predicate(mol, -1)));
  ASSERT( molecule_has_atom(mol, charge_lt_predicate(mol, 2)));
  ASSERT( molecule_has_atom(mol, charge_lt_predicate(mol, 1)));
  ASSERT(!molecule_has_atom(mol, charge_lt_predicate(mol, 0)));
  ASSERT( molecule_has_atom(mol, charge_gt_predicate(mol, 0)));
  ASSERT(!molecule_has_atom(mol, charge_gt_predicate(mol, 1)));
  ASSERT( molecule_has_atom(mol, charge_leq_predicate(mol, 1)));
  ASSERT( molecule_has_atom(mol, charge_leq_predicate(mol, 0)));
  ASSERT(!molecule_has_atom(mol, charge_leq_predicate(mol, -1)));
  ASSERT( molecule_has_atom(mol, charge_geq_predicate(mol, 1)));
  ASSERT(!molecule_has_atom(mol, charge_geq_predicate(mol, 2)));
}

void test_degree_predicate()
{
  // DegreePredicate
  HeMol mol = hemol_from_smiles("CC(O)C");
  ASSERT( molecule_has_atom(mol, degree_eq_predicate(mol, 1)));
  ASSERT(!molecule_has_atom(mol, degree_eq_predicate(mol, 2)));
  ASSERT( molecule_has_atom(mol, degree_eq_predicate(mol, 3)));
  ASSERT( molecule_has_atom(mol, degree_lt_predicate(mol, 2)));
  ASSERT(!molecule_has_atom(mol, degree_lt_predicate(mol, 1)));
  ASSERT( molecule_has_atom(mol, degree_gt_predicate(mol, 2)));
  ASSERT(!molecule_has_atom(mol, degree_gt_predicate(mol, 3)));
  ASSERT( molecule_has_atom(mol, degree_leq_predicate(mol, 1)));
  ASSERT(!molecule_has_atom(mol, degree_leq_predicate(mol, 0)));
  ASSERT( molecule_has_atom(mol, degree_geq_predicate(mol, 3)));
  ASSERT(!molecule_has_atom(mol, degree_geq_predicate(mol, 4)));
}

void test_aromatic_atom_predicate()
{
  // AromaticAtomPredicate
  HeMol mol1 = hemol_from_smiles("c1ccccc1");
  ASSERT( molecule_has_atom(mol1, aromatic_atom_predicate(mol1)));
  ASSERT(!molecule_has_atom(mol1, not_aromatic_atom_predicate(mol1)));

  HeMol mol2 = hemol_from_smiles("C1CCCCC1");
  ASSERT(!molecule_has_atom(mol2, aromatic_atom_predicate(mol2)));
  ASSERT( molecule_has_atom(mol2, not_aromatic_atom_predicate(mol2)));
}

void test_atom_and_predicate()
{
  HeMol mol = hemol_from_smiles("[13C]");

  ASSERT( atom_and_predicates(mol, element_eq_predicate(mol, 6), mass_eq_predicate(mol, 13))(mol, get_atom(mol, 0)));
  ASSERT(!atom_and_predicates(mol, element_eq_predicate(mol, 7), mass_eq_predicate(mol, 13))(mol, get_atom(mol, 0)));
  ASSERT(!atom_and_predicates(mol, element_eq_predicate(mol, 6), mass_eq_predicate(mol, 14))(mol, get_atom(mol, 0)));
}

void test_atom_or_predicate()
{
  HeMol mol = hemol_from_smiles("[13C]");

  ASSERT( atom_or_predicates(mol, element_eq_predicate(mol, 6), mass_eq_predicate(mol, 13))(mol, get_atom(mol, 0)));
  ASSERT( atom_or_predicates(mol, element_eq_predicate(mol, 7), mass_eq_predicate(mol, 13))(mol, get_atom(mol, 0)));
  ASSERT( atom_or_predicates(mol, element_eq_predicate(mol, 6), mass_eq_predicate(mol, 14))(mol, get_atom(mol, 0)));
  ASSERT(!atom_or_predicates(mol, element_eq_predicate(mol, 7), mass_eq_predicate(mol, 14))(mol, get_atom(mol, 0)));
}

void test_atom_predicates()
{
  test_element_predicate();
  test_mass_predicate();
  test_charge_predicate();
  test_degree_predicate();
  test_aromatic_atom_predicate();
  test_atom_and_predicate();
  test_atom_or_predicate();
}

void test_molecule_has_atom()
{
  HeMol methane = hemol_from_smiles("C");

  ASSERT(molecule_has_atom(methane, ElementPredicate<HeMol>(6)));
  ASSERT(!molecule_has_atom(methane, ElementPredicate<HeMol>(7)));
}

template<typename MoleculeType>
void test_editable_molecule()
{
  typedef typename molecule_traits<MoleculeType>::atom_type atom_type;
  typedef typename molecule_traits<MoleculeType>::bond_type bond_type;
  MoleculeType mol;

  // create 4 atoms
  atom_type atom1 = add_atom(mol);
  COMPARE(1, num_atoms(mol));
  atom_type atom2 = add_atom(mol);
  COMPARE(2, num_atoms(mol));
  atom_type atom3 = add_atom(mol);
  COMPARE(3, num_atoms(mol));
  atom_type atom4 = add_atom(mol);
  COMPARE(4, num_atoms(mol));

  // set some atom properties
  set_aromatic(mol, atom1, true);
  set_aromatic(mol, atom2, true);
  set_aromatic(mol, atom3, false);
  set_aromatic(mol, atom4, false);
  COMPARE(true, is_aromatic(mol, atom1));
  COMPARE(true, is_aromatic(mol, atom2));
  COMPARE(false, is_aromatic(mol, atom3));
  COMPARE(false, is_aromatic(mol, atom4));

  set_element(mol, atom1, 6);
  set_element(mol, atom2, 7);
  set_element(mol, atom3, 8);
  set_element(mol, atom4, 9);
  COMPARE(6, get_element(mol, atom1));
  COMPARE(7, get_element(mol, atom2));
  COMPARE(8, get_element(mol, atom3));
  COMPARE(9, get_element(mol, atom4));

  set_mass(mol, atom1, 60);
  set_mass(mol, atom2, 70);
  set_mass(mol, atom3, 80);
  set_mass(mol, atom4, 90);
  COMPARE(60, get_mass(mol, atom1));
  COMPARE(70, get_mass(mol, atom2));
  COMPARE(80, get_mass(mol, atom3));
  COMPARE(90, get_mass(mol, atom4));

  set_hydrogens(mol, atom1, 1);
  set_hydrogens(mol, atom2, 2);
  set_hydrogens(mol, atom3, 3);
  set_hydrogens(mol, atom4, 4);
  COMPARE(1, get_hydrogens(mol, atom1));
  COMPARE(2, get_hydrogens(mol, atom2));
  COMPARE(3, get_hydrogens(mol, atom3));
  COMPARE(4, get_hydrogens(mol, atom4));

  set_charge(mol, atom1, -2);
  set_charge(mol, atom2, -1);
  set_charge(mol, atom3, 0);
  set_charge(mol, atom4, 2);
  COMPARE(-2, get_charge(mol, atom1));
  COMPARE(-1, get_charge(mol, atom2));
  COMPARE(0, get_charge(mol, atom3));
  COMPARE(2, get_charge(mol, atom4));

  // add 3 bonds
  bond_type bond1 = add_bond(mol, atom1, atom2);
  COMPARE(1, num_bonds(mol));
  bond_type bond2 = add_bond(mol, atom2, atom3);
  COMPARE(2, num_bonds(mol));
  bond_type bond3 = add_bond(mol, atom3, atom4);
  COMPARE(3, num_bonds(mol));

  // set some bond properties
  set_aromatic(mol, bond1, false);
  set_aromatic(mol, bond2, true);
  set_aromatic(mol, bond3, false);
  COMPARE(false, is_aromatic(mol, bond1));
  COMPARE(true, is_aromatic(mol, bond2));
  COMPARE(false, is_aromatic(mol, bond3));

  set_order(mol, bond1, 1);
  set_order(mol, bond2, 2);
  set_order(mol, bond3, 3);
  COMPARE(1, get_order(mol, bond1));
  COMPARE(2, get_order(mol, bond2));
  COMPARE(3, get_order(mol, bond3));

  // remove a bond
  remove_bond(mol, bond3);
  COMPARE(4, num_atoms(mol));
  COMPARE(2, num_bonds(mol));
  COMPARE(1, get_order(mol, get_bond(mol, 0)));
  COMPARE(2, get_order(mol, get_bond(mol, 1)));

  // remove an atom
  remove_atom(mol, atom2);
  COMPARE(3, num_atoms(mol));
  COMPARE(0, num_bonds(mol));
  COMPARE(60, get_mass(mol, get_atom(mol, 0)));
  COMPARE(80, get_mass(mol, get_atom(mol, 1)));
  COMPARE(90, get_mass(mol, get_atom(mol, 2)));

  // clear the molecule
  clear_molecule(mol);
  COMPARE(0, num_atoms(mol));
  COMPARE(0, num_bonds(mol));
}

void test_substructure()
{
  HeMol mol = hemol_from_smiles("Oc1cc(CC)ccc1N");

  std::vector<bool> atoms(mol.numAtoms(), true);
  atoms[0] = false;
  atoms[4] = false;
  atoms[5] = false;
  atoms[9] = false;

  std::vector<bool> bonds(mol.numBonds(), true);
  bonds[0] = false;
  bonds[3] = false;
  bonds[4] = false;
  bonds[9] = false;

  HeMol sub;
  make_substructure(sub, mol, atoms, bonds);

  Smiles SMILES;
  COMPARE("c1ccccc1", SMILES.write(sub));
}

int main()
{
  test_substructure();

  test_atom_predicates();
  test_molecule_has_atom();

  test_editable_molecule<HeMol>();
  test_editable_molecule<SmartMol>();

  HeMol hemol;
  //check_editable_molecule_concept(hemol);

  /*
  try {
    SmartMol smartmol;
    check_editable_molecule_concept(smartmol);
  } catch (...) {}
  */
}

File: test/python/molecule.py
import helium
import unittest

SMILES = helium.Smiles()

class TestMolecule(unittest.TestCase):

    def test_molecule(self):
        mol = helium.Molecule()
        SMILES.read('CC(C)N', mol)

        self.assertEqual(4, mol.numAtoms())
        self.assertEqual(3, mol.numBonds())

        atoms = []
        for atom in mol.atoms():
            atoms.append(atom)
        self.assertListEqual([mol.atom(0), mol.atom(1), mol.atom(2), mol.atom(3)], atoms)

        bonds = []
        for bond in mol.bonds():
            bonds.append(bond)
        self.assertListEqual([mol.bond(0), mol.bond(1), mol.bond(2)], bonds)

    def test_atom(self):
        mol = helium.Molecule()
        SMILES.read('CC(C)N', mol)

        atom1 = mol.atom(0)
        atom2 = mol.atom(1)
        atom3 = mol.atom(2)
        atom4 = mol.atom(3)

        self.assertEqual(0, atom1.index())
        self.assertEqual(1, atom2.index())
        self.assertEqual(2, atom3.index())
        self.assertEqual(3, atom4.index())

        self.assertEqual(6, atom1.element())
        self.assertEqual(6, atom2.element())
        self.assertEqual(6, atom3.element())
        self.assertEqual(7, atom4.element())

        self.assertEqual(1, atom1.degree())
        self.assertEqual(3, atom2.degree())
        self.assertEqual(1, atom3.degree())
        self.assertEqual(1, atom4.degree())

        indices = []
        for nbr in atom2.nbrs():
            indices.append(nbr.index())
        self.assertListEqual([0, 2, 3], indices)

        indices = []
        for bond in atom2.bonds():
            indices.append(bond.index())
        self.assertListEqual([0, 1, 2], indices)

    def test_bond(self):
        mol = helium.Molecule()
        SMILES.read('CC(C)=O', mol)

        atom1 = mol.atom(0)
        atom2 = mol.atom(1)
        atom3 = mol.atom(2)
        atom4 = mol.atom(3)

        bond1 = mol.bond(0)
        bond2 = mol.bond(1)
        bond3 = mol.bond(2)

        self.assertEqual(0, bond1.index())
        self.assertEqual(1, bond2.index())
        self.assertEqual(2, bond3.index())

        self.assertEqual(1, bond1.order())
        self.assertEqual(1, bond2.order())
        self.assertEqual(2, bond3.order())

        self.assertEqual(atom1, bond1.source())
        self.assertEqual(atom2, bond2.source())
        self.assertEqual(atom2, bond3.source())

        self.assertEqual(atom2, bond1.target())
        self.assertEqual(atom3, bond2.target())
        self.assertEqual(atom4, bond3.target())

        self.assertEqual(atom1, bond1.other(atom2))
        self.assertEqual(atom2, bond2.other(atom3))
        self.assertEqual(atom2, bond3.other(atom4))

    def test_atom_editing(self):
        mol = helium.Molecule()
        atom = mol.addAtom()

        self.assertEqual(0, atom.index())

        atom.setElement(6)
        self.assertEqual(6, atom.element())
        atom.setElement(7)
        self.assertEqual(7, atom.element())

        atom.setAromatic(True)
        self.assertTrue(atom.isAromatic())
        atom.setAromatic(False)
        self.assertFalse(atom.isAromatic())

        atom.setMass(12)
        self.assertEqual(12, atom.mass())
        atom.setMass(13)
        self.assertEqual(13, atom.mass())

        atom.setHydrogens(4)
        self.assertEqual(4, atom.hydrogens())
        atom.setHydrogens(3)
        self.assertEqual(3, atom.hydrogens())

        atom.setCharge(1)
        self.assertEqual(1, atom.charge())
        atom.setCharge(-1)
        self.assertEqual(-1, atom.charge())

    def test_bond_editing(self):
        mol = helium.Molecule()
        source = mol.addAtom()
        target = mol.addAtom()
        bond = mol.addBond(source, target)

        bond.setAromatic(True)
        self.assertTrue(bond.isAromatic())
        bond.setAromatic(False)
        self.assertFalse(bond.isAromatic())

        bond.setOrder(1)
        self.assertEqual(1, bond.order())
        bond.setOrder(2)
        self.assertEqual(2, bond.order())

    def test_molecule_editing(self):
        mol = helium.Molecule()

        atom1 = mol.addAtom()
        atom2 = mol.addAtom()
        atom3 = mol.addAtom()
        atom4 = mol.addAtom()

        bond1 = mol.addBond(atom1, atom2)
        bond2 = mol.addBond(atom2, atom3)
        bond3 = mol.addBond(atom3, atom4)

        self.assertEqual(4, mol.numAtoms())
        self.assertEqual(3, mol.numBonds())

        mol.removeBond(bond3)

        self.assertEqual(4, mol.numAtoms())
        self.assertEqual(2, mol.numBonds())

        mol.removeAtom(atom1)

        self.assertEqual(3, mol.numAtoms())
        self.assertEqual(1, mol.numBonds())

        mol.clear()

        self.assertEqual(0, mol.numAtoms())
        self.assertEqual(0, mol.numBonds())

    def test_copy_ctor(self):
        mol1 = helium.Molecule()

        SMILES.read('CC', mol1)

        mol2 = helium.Molecule(mol1)

        self.assertEqual(2, mol1.numAtoms())
        self.assertEqual(2, mol2.numAtoms())

        mol2.clear()

        self.assertEqual(2, mol1.numAtoms())
        self.assertEqual(0, mol2.numAtoms())

    def test_get_bond(self):
        mol = helium.Molecule()
        SMILES.read('C=C', mol)

        bond = mol.bond(0)
        bond = mol.bond(mol.atom(0), mol.atom(1))

if __name__ == '__main__':
    unittest.main()
